File: java/78.subsets.java
/*
 * @lc app=leetcode id=78 lang=java
 *
 * [78] Subsets
 */

// @lc code=start
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        Deque<Integer> q = new ArrayDeque();
        List<List<Integer>> res = new ArrayList();

        buildSubset(nums, res, q, -1);

        return res;
    }

    public void buildSubset(int[] nums, List<List<Integer>> res, Deque<Integer> q, int i) {
        res.add(new ArrayList(q));

        for(int j = i + 1; j < nums.length; j++) {
            q.addLast(nums[j]);
            buildSubset(nums, res, q, j);
            q.removeLast();
        }
    }
}
// @lc code=end

File: cpp/78.subsets.cpp
/*
 * @lc app=leetcode id=78 lang=cpp
 *
 * [78] Subsets
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> curr;
        subset(res, nums, curr, 0);
        
        return res;
    }

    void subset(vector<vector<int>>& res, vector<int>& nums, vector<int>& curr, int i) {
        res.push_back(curr);

        for(int j = i; j < nums.size(); j++) {
            curr.push_back(nums[j]);
            subset(res, nums, curr, j + 1);
            curr.pop_back();
        }
    }
};
// @lc code=end

